spark/src/main/scala/geotrellis/spark/io/tms/TMSReader.scala

```diff
@@ -0,0 +1,84 @@
+package geotrellis.spark.io.tms
+
+import geotrellis.raster._
+import geotrellis.raster.render._
+import geotrellis.spark.SpatialKey
+import geotrellis.vector.{Point, Extent}
+import geotrellis.spark.tiling.ZoomedLayoutScheme
+
+import spire.syntax.cfor._
+import scalaj.http._
+
+import scala.collection.mutable._
+import java.net.{URI, URL}
+import javax.imageio._
+
+
+class TMSReader[T](
+  uriTemplate: String,
+  f: URI => T = TMSReader.decodeTile _
+) extends Serializable {
+  def uri(z: Int, x: Int, y: Int): URI = new URI(
+    uriTemplate.replace("{z}", z.toString)
+      .replace("{x}", x.toString)
+      .replace("{y}", y.toString)
+  )
+
+  def read(zoom: Int, bounds: GridBounds): Iterator[(SpatialKey, T)] = {
+    // construct bounds from zoom + extent
+    for ((x, y) <- bounds.coordsIter)
+    yield SpatialKey(x, y) -> f(uri(zoom, x, y))
+  }
+
+  def read(zoom: Int, extent: Extent, tileSize: Int = ZoomedLayoutScheme.DEFAULT_TILE_SIZE): Iterator[(SpatialKey, T)] = {
+    val layoutDefinition = ZoomedLayoutScheme.layoutForZoom(zoom, extent, tileSize)
+    val bounds = layoutDefinition.mapTransform.extentToBounds(extent)
+    read(zoom, bounds)
+  }
+
+  def read(zoom: Int, x: Int, y: Int): T = f(uri(zoom, x, y))
+}
+
+object TMSReader {
+
+  def apply[T](uriTemplate: String, f: URI => T = TMSReader.decodeTile _): TMSReader[T] =
+    new TMSReader(uriTemplate, f)
+
+  // ImageIO use here to deal with different decoding behaviors (png vs jpg)
+  def decodeTile(uri: URI): Tile = {
+    val img = ImageIO.read(uri.toURL)
```

Comment:
It's a `png` / `jpg` reader only?

Reply:
Is any other format supported by TMS?
http://mapserver.org/mapcache/formats.html

Reply:
@metasim that's what i wanted to see actually :D thanks for reference!